contracts/CurrencyNetwork.sol

```diff
@@ -844,74 +814,61 @@ contract CurrencyNetwork is CurrencyNetworkInterface, Authorizable, debtTracking
 
        this function receives the path along which to do the transfer. This path
        is computed by the relay server based on the then current state of the
-       balance. In case the balance changed it's sign, the path will not have
+       balance. In case the balance changed its sign, the path will not have
        the right 'shape' and the require statements below will revert the
        transaction.
 
-       XXX This function is currently broken for balances which do not fit into
+       TODO: This function is currently broken for balances which do not fit into
        a uint32. We may repair that later when merging the interest changes.
      */
     function _closeTrustlineByTriangularTransfer(
-        address _from,
         address _otherParty,
         uint32 _maxFee,
         address[] memory _path)
         internal
     {
-        Trustline memory trustline = _loadTrustline(_from, _otherParty);
+        require(
+            _path.length >= 3,
+            "Path given too short to be correct."
+        );
+        require(msg.sender == _path[0], "Path must start msg.sender.");
+        require(
+            msg.sender == _path[_path.length - 1],
+            "Last element of path does not match msg.sender."
+        );
+
+        Trustline memory trustline = _loadTrustline(_path[0], _otherParty);
         require(!_isTrustlineFrozen(trustline.agreement), "The trustline is frozen and cannot be closed.");
         _applyInterests(trustline);
-        /* we could as well call _storeTrustlineBalances here. It doesn't matter for the
-           _mediatedTransfer/_mediatedTransferReceiverPays calls below since the
-           interest will be recomputed if we don't call storeTrustlineBalances here. We
-           may investigate what's cheaper gas-wise later.
-        */
         TrustlineBalances memory balances = trustline.balances;
+
         if (balances.balance > 0) {
             require(
-                _path.length >= 2,
-                "The path given is too short to be correct."
-            );
-            require(
-                _from == _path[_path.length - 1],
-                "The last element of the path does not match with the _from address."
-            );
-            require(
-                _path[0] == _otherParty,
-                "The first element of the path does not match with the _otherParty address."
+                _path[1] == _otherParty,
```

Comment:
do we need `otherParty` and why?

Reply:
Information theoretically we do not need it but the alternatives are not so much better.

If you have trustlines in a triangle: `A <-> B <-> C <-> A` and `A` wants to close the trustline with either B or C, the `path` he will use depends on the unbalance of the network and not on  whether he wants to close his trustline with B or C.

So to determine whether we need to close the `A <-> B` or `A <-> C` trustline we need additional information.

We could remove the counterparty arg and assume that the counterparty is always the second element of path and then determine in the contract that the transfer should be in reverse order of the path when the balance of the TL is negative. But I feel like this is more confusing (and probably more gas expensive).

Otherwise we could have a bool that indicates whether the counterparty is second element or second to last element of the path. I do not think this is strictly better.